static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,152 @@
+
+Repositories containing all our code and schematics:
+* esp32c3 code: [https://github.com/rustbox/esp32c3-vgaterm](https://github.com/rustbox/esp32c3-vgaterm)
+* VGA timing generation code and Programmable Logic Device code: [https://github.com/rustbox/cuddly-robot](https://github.com/rustbox/cuddly-robot)
+* PCB schematics and CAD: [https://github.com/rustbox/vgaterm-schematics](https://github.com/rustbox/vgaterm-schematics)
+
+### VGA Timing
+
+VGA comes in several modes, and we target the 640x480, 60Hz frame rate mode - the base standard. VGA has 3 analog signals between 0 and 0.7 volts representing the amount of Red Green and Blue per pixel. There are also two timing signals `H_SYNC` and `V_SYNC` that tell the monitor what resolution mode to be in and when to sample the analog color signals to actually dispaly a pixel. When the monitor is actively sampling colors and displaying pixels we say that this portion of the frame is `visible`. There is time during a complete frame when this isn't the case and this is _blanking area_. Lastly, the _pixel clock_ speed is the rate at which pixels are shown, and in our case that's 25.175 MHz. In the diagram above, I've labeled a signal `CLK_VIS`, and this is the _pixel clock_ signal during the `visible` time. To see detailed VGA timing information, see [tiny vga](http://tinyvga.com/vga-timing/640x480@60Hz).
+
+:----:|:-----:
+![](/assets/images/vgaterm-timing-front.jpg) | ![](/assets/images/vgaterm-timing-back.jpg)
+<span style="color: {{ site.caption_color }};">Vgaterm Timing circuit top</span> | <span style="color: {{ site.caption_color }};">Vgaterm Timing cicuit bottom</span>
+
+VGA timing is controlled by the "timing board" circuit. This establishes the `H_SYNC` and `V_SYNC` signals that VGA requires, the `START` signal which alerts the CPU to begin  emitting pixels to the FIFO, and the `visible` flag, which is used to compute `CLK_VIS` to read pixel bytes out of the FIFO.
+
+### Pixels Out
+Every frame (60 times a second) the esp32c3 microcontroller needs to output 256,000 bytes (640x400 pixels, 1 byte per pixel) to hardware so they may be displayed. This is the most important software role as any jitter in timing can cause display artifacts.
+
+To cross the clock domains, from the CPU SPI to VGA pixel clock, we used a FIFO memory chip. This lets us write bytes from the CPU while VGA hardware reads bytes exactly when it needs them.
+
+This was clever, but there are a few constraints we had to consider when designing a circuit around the FIFO:
+* The FIFO read and write setup and hold times had to be obeyed by the SPI clock and `CLK_VIS` signals
+* The SPI Clock could be in only divisions by 2 of 80 MHz (80, 40, 20, 10, etc)
+* If we emitted pixels too quickly, and VGA hardware could not read from the FIFO fast enough, we would blow the queue and drop pixels
+    * This could be mitigated by adding more space in the FIFO, but at a steep dollar cost
```

Comment:
```suggestion
```

(and cover the dollars-to-safety-margin tradeoff earlier)

Reply:
I was thinking the first section where you had the comment about cost was to just introduce everything and this section would go into a little more detail where the cost would be talked about along with the other constraints

Reply:
It's just that this was the only sub-bullet in the list, and I don't know that it justified its existence. Especially given it's probably worth adding this much context when we're introducing the FIFO, which happened earlier in the "Technical Overview".

If you did want to go more into detail here, that would be great: I'd love to see e.g. "the FIFO was $20, which was 5-20x all our other ICs, so we only had a couple from when we just wild-ass guessed that 4k would be a good size, and we didn't want to buy bigger ones" or some sort of a  $/µs for a 4k vs 8k FIFO retroactive justification analysis